### PR TITLE
Improve Program Defensiveness

### LINKED ISSUE #281: Duplicate options bug
Specifying the same option more than once in user commands is silently accepted, and seems to affect program behaviour in unexpected ways. For instance running
``` console
binbash> add -n something -d desc -c 5 -n these -re
```
results in:
> Noted! I have added the following item into your inventory:
>
> [R] something these
>         description: desc
>         quantity: 0
>         cost price: $5.00
>         sale price: $0.00
>         threshold: 1
> -------------------------------------------------------------

### LINKED ISSUE #282: Empty string invalid command bug
Giving the empty string by hitting `enter` on a blankline or whitespace seems to result in the "invalid command" error message. This might not be the ideal way to handle such inputs.

### PR DESCRIPTION
**Pull request by nkotaa**
# Description
This PR aims to generally improve the defensiveness of our program. Namely the following changes have been made:
1. Fix #281 
2. Fix #282 
3. Handle program EOF and interrupts more meaningfully
4. Improve test coverage for parsers
5. Improve defensiveness of `Parser`, `Ui` and `CommandOptionAdder`
6. Update UG sections towards #59 
7. [Enhancement] Exception messages printed with different format

### CODE REVIEW

File: src/main/java/seedu/binbash/parser/Parser.java [hunk 1]
```diff
@@ -135,6 +138,23 @@ private ListCommand parseListCommand(String[] commandArgs) throws InvalidCommand
...
+     *
+     * @param processedOptions The list of options processed.
+     * @throws ParseException If duplicate option found.
+     */
+    static void checkDuplicateOption(Option[] processedOptions) throws ParseException {
+        HashSet<String> optionSet = new HashSet<>();
+        for (Option processedOption : processedOptions) {
+            String currentOption = processedOption.getOpt();
+            if (optionSet.contains(currentOption)) {
+                throw new ParseException("Duplicate option: -" + currentOption);
```
Comment (nur-haziq) [inline src/main/java/seedu/binbash/parser/Parser.java:152]: Good use of HashSet 👍

File: src/main/java/seedu/binbash/parser/SearchCommandParser.java [hunk 2]
```diff
@@ -148,4 +105,67 @@ String[] parseRangeArgument(String argument, String option) throws ParseExceptio
...
+            dateRange[0] = Parser.parseDateOptionValue(rangeArgument[0], "expiry date lower bound");
+        }
+        if (rangeArgument[1] != "") {
+            dateRange[1] = Parser.parseDateOptionValue(rangeArgument[1], "expiry date upper bound");
+        }
+        if (dateRange[0].isAfter(dateRange[1])) {
+            throw new ParseException(type + " lower bound is more than upper bound");
+        }
+        return dateRange;
+    }
```
Comment (nur-haziq) [inline src/main/java/seedu/binbash/parser/SearchCommandParser.java:161]: Perhaps 0 and 1 can be renamed to LOWER_BOUND and UPPER_BOUND respectively to reduce use of magic literals, but this is extremely minor. Code is still clean and readable. 👍

File: src/main/java/seedu/binbash/ui/Ui.java [hunk 3]
```diff
@@ -86,11 +90,33 @@ public void greet() {
...
+     * Prints a warning of some unexpected application behaviour to standard output.
+     *
+     * @param line The text to print.
+     */
+    public void warn(String line) {
+        if (!isUserActive) {
+            assert line == "";
+        }
+        System.out.println(ANSI_RED + line + ANSI_RESET);
+    }
```
Comment (nur-haziq) [inline src/main/java/seedu/binbash/ui/Ui.java:114]: Good implementation to distinguish between a `talk()` output and a `warn()` output.

**Review by nur-haziq — APPROVED** [review]
LGTM. Thanks for improving the defensiveness of our Parser, and for bolstering our test coverage. 😄

File: docs/UserGuide.md
Comment (YHWong20) [inline docs/UserGuide.md:1]: Thanks for helping to update this section!

File: docs/diagrams/ParserClassDiagram.puml [hunk 4]
```diff
@@ -6,6 +6,10 @@ skinparam CircledCharacterRadius 0
 
 class Parser {
     + parseCommand(userInput: String): Command
+    ~ checkDuplicateOption(processedOptions: Option[]) {static}
+    ~ parseIntOptionValue(argument: String, option: String): int {static}
+    ~ parseDoubleOptionValue(argument: String, option: String): double {static}
+    ~ parseDateOptionValue(argument: String, option: String): LocalDate {static}
```
Comment (YHWong20) [inline docs/diagrams/ParserClassDiagram.puml:12]: As the `checkDuplicateOption()` method/functionality is new, should we include this in the DG as well? Perhaps just a mention that the Parser can recognize the presence of duplicate options.

Also, I'm not sure if we need to include the `parseXYZOptionValue()` methods in here, as they are helper methods.

File: src/main/java/seedu/binbash/parser/SearchCommandParser.java (hunk 2 above)
Comment (nkotaa) [reply src/main/java/seedu/binbash/parser/SearchCommandParser.java]: You have an eye for magic literals. Removed as suggested.

File: docs/diagrams/ParserClassDiagram.puml (hunk 4 above)
Comment (nkotaa) [reply docs/diagrams/ParserClassDiagram.puml:12]: DG explanation has been updated to give more relevance to diagram.

**Pull request merged**